quinn-h3/src/frame.rs

```diff
@@ -1,519 +1,167 @@
-            buf.write_var(dependency)
+            let size = match self.recv.poll_read(&mut self.buf[start..])? {
+                Async::NotReady => 0,
+                Async::Ready(size) => {
+                    if size == 0 {
+                        self.finished = true;
+                    }
+                    size
+                }
+            };
+            self.buf.truncate(self.buf.len() - Self::READ_SIZE + size);
         }
-        buf.write(self.weight);
-    }
-}
 
-impl FrameHeader for PriorityFrame {
-    const TYPE: Type = Type::PRIORITY;
-    fn len(&self) -> usize {
-        let mut size = size_of::<u8>() * 2;
-
-        size += match self.prioritized {
-            Priority::RequestStream(id) | Priority::PushStream(id) | Priority::Placeholder(id) => {
-                varint::size(id).unwrap()
-            }
-            _ => 0,
+        let (pos, decoded) = {
+            let mut cur = io::Cursor::new(&mut self.buf);
+            let decoded = HttpFrame::decode(&mut cur);
+            (cur.position() as usize, decoded)
         };
-        size += match self.dependency {
-            Priority::RequestStream(id) | Priority::PushStream(id) | Priority::Placeholder(id) => {
-                varint::size(id).unwrap()
-            }
-            _ => 0,
-        };
-
-        size
-    }
-}
 
-#[derive(Debug, PartialEq)]
-pub struct SettingsFrame {
-    pub num_placeholders: u64,
-    pub max_header_list_size: u64,
-}
-
-impl Default for SettingsFrame {
-    fn default() -> SettingsFrame {
-        SettingsFrame {
-            num_placeholders: 16,
-            max_header_list_size: 65536,
-        }
-    }
-}
-
-impl SettingsFrame {
-    fn encode<T: BufMut>(&self, buf: &mut T) {
-        self.encode_header(buf);
-        SettingId::NUM_PLACEHOLDERS.encode(buf);
-        buf.write_var(self.num_placeholders);
-        SettingId::MAX_HEADER_LIST_SIZE.encode(buf);
-        buf.write_var(self.max_header_list_size);
-    }
-
-    fn decode<T: Buf>(buf: &mut T) -> Result<SettingsFrame, Error> {
-        let mut settings = SettingsFrame::default();
-        while buf.has_remaining() {
-            if buf.remaining() < 2 {
-                // remains less than 2 * minimum-size varint
-                return Err(Error::Malformed);
-            }
-            let identifier = SettingId::decode(buf).map_err(|_| Error::Malformed)?;
-            let value = buf.get_var().map_err(|_| Error::InvalidFrameValue)?;
-            match identifier {
-                id if id.0 & 0x0f0f == 0x0a0a => continue,
-                SettingId::NUM_PLACEHOLDERS => {
-                    settings.num_placeholders = value;
-                }
-                SettingId::MAX_HEADER_LIST_SIZE => {
-                    settings.max_header_list_size = value;
+        return match decoded {
+            Err(frame::Error::UnexpectedEnd) => match self.buf.len() {
+                Self::READ_SIZE => Err(Error::Overflow),
```

Comment:
This won't play well with big `Data` frames. Either the buffer should grow, either a partial frame decoding should be implemented, to let the user of this type stream chunks somehow.

I decided to leave this as is for now, looking forward to body streaming feature for the end user.